### PR TITLE
⚠️ Update tests to Ginkgo v2

### LINKED ISSUE #4897: Replatform e2e on to ginkgo v2
**User Story**

As a developer, I want my infrastructure cleaned up properly after an aborted test run.

**Detailed Description**

Looks like Ginkgo v2 will solve a lot of problems when it comes to testing on actual infrastructure. One of the main benefits is that `AfterEach` will be executed when tests are aborted, whereas for now only `AfterSuite` is. In AWS, with the current behaviour, this leaves a lot of resources behind when consuming the CAPI test framework, because we don't know the namespace the framework has created which we normally store in a node-level map for the AfterSuite to iterate through and clean up clusters.

As replatforming onto v2 (which isn't released yet) will be a breaking change, very much a v0.5.0 thing.

/milestone next

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind feature
/area testing
/priority important-longterm

### PR DESCRIPTION
**Pull request by mboersma**
**What this PR does / why we need it**:

Updates CAPI to [Ginkgo v2](https://onsi.github.io/ginkgo/MIGRATING_TO_V2#upgrading-to-ginkgo-20).

Kubernetes just migrated in kubernetes/kubernetes#109111. I attempted to update Ginkgo [in CAPZ](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2466) but can't because of test framework depencies here in CAPI.

The reporting framework has changed significantly: essentially, custom reporters are out. The `--junit-report` and related flags *should* give us similar output files, but let's see.

**Which issue(s) this PR fixes**:

Fixes #4897

cc: @vladimirvivien

### CODE REVIEW

File: go.mod
Comment (sbueringer) [inline go.mod:24]: Usually we keep those versions in sync with CR. (https://github.com/kubernetes-sigs/controller-runtime/blob/master/go.mod)

Should we make an exception in this case? 

I think we don't really depend on using the same ginkgo version as CR. It's also a /v2 dependency so from
a go module perspective it's kind of another dependency and we could still important the same v1 as CR if that becomes necessary.

@vincepri WDYT?

Comment (jackfrancis) [reply go.mod:24]: If we update ginkgo to v2 in capz @ main then we won't be able to import capi 1.2 without this change landing in 1.2

Comment (mboersma) [reply go.mod:24]: > Should we make an exception in this case?

I'm not necessarily arguing for that. Maybe the right thing to do is update controller-runtime to Ginkgo v2 first.

Comment (mboersma) [reply go.mod:24]: > If we update ginkgo to v2 in capz @ main...

I think updating Ginkgo in provider projects is effectively blocked until this merges. You can't import v1 and v2 at the same time since they'll conflict in handling [flags](https://github.com/onsi/ginkgo/issues/875#issuecomment-1036350659).

Comment (sbueringer) [reply go.mod:24]: Forgot: cc @fabriziopandini

File: test/e2e/Makefile
Comment (fabriziopandini) [inline test/e2e/Makefile:129]: Q: The 01 in `junit.e2e_suite.01.xml` previously was the number of the node, and we got more files, one for each node written in parallel. How does this works now?

**Comment by mboersma** [comment]
/retitle [WIP] Update tests to Ginkgo v2

**Comment by enxebre** [comment]
>I think we should only do this if there is a good reason why we have to use ginkgo v2 in CAPI v1.2

+1, it does not seem to be the case at the moment.

File: test/e2e/Makefile
Comment (mboersma) [reply test/e2e/Makefile]: Apparently `--output-dir` is effectively required when using `--junit-report`, since the latter doesn't handle absolute paths. I'm not the only one who found this [confusing](https://github.com/onsi/ginkgo/issues/1001).

**Comment by CecileRobertMichon** [comment]
Looks like it's counting SynchronizedBeforeSuite as a spec in the Junit reporting, not sure if that's expected or not

Screen Shot 2022-07-20 at 3 53 12 PM

**Comment by sbueringer** [comment]
I should have time for a review early next week.

**Comment by sbueringer** [comment]
cc @schrej 
If you have some time can you take a look at this PR as well, given your experience working on the Ginkgo v2 bump in CR>

**Comment by schrej** [comment]
Switching imports and replacing the custom JUnit reporter with command line flags is what we've done on CR as well, so 
/lgtm

Should maybe be a :warning: in the title though?

**Review by killianmuldoon — COMMENTED** [review]
Should we add a note to the 1.2->1.3 guide mentioning this?

**Comment by sbueringer** [comment]
> Should we add a note to the 1.2->1.3 guide mentioning this?

Yes. We should definitely mention:
* that the ginkgo & gomega dependencies were upgraded and thus all consumers have to ugprade as well
* the default timeout has been changed to 1h (https://onsi.github.io/ginkgo/MIGRATING_TO_V2#timeout-behavior)
* -juni-report has to be used instead of the junit reporter
* xref to this PR for an example-upgrade & to https://onsi.github.io/ginkgo/MIGRATING_TO_V2#timeout-behavior

**Comment by sbueringer** [comment]
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-1-22-main
/test pull-cluster-api-e2e-workload-upgrade-1-24-latest-main

Want to see if there are any side effects on the upgrade jobs which are running the upstream conformance tests

**Comment by mboersma** [comment]
> Should we add a note to the 1.2->1.3 guide mentioning this?

👍🏻 I see that we have mentioned test framework changes before under the `Other:` section in version migration docs, so I'll add that.

**Comment by mboersma** [comment]
/retitle ⚠️ Update tests to Ginkgo v2

**Review by killianmuldoon — APPROVED** [review]
/lgtm

**Comment by sbueringer** [comment]
Thank you!

lgtm pending green tests

**Comment by sbueringer** [comment]
I assume otherwise we're basically ready to merge (considering the mail wrote we want to merge today :)).

cc @CecileRobertMichon @fabriziopandini

**Comment by mboersma** [comment]
> Looks like it's counting SynchronizedBeforeSuite as a spec in the Junit reporting, not sure if that's expected or not

@CecileRobertMichon I see that changed, but I haven't been able to track down where and why. I assume it's part of the new `--junit-report` implementation. It seems benign, but I can look into it more deeply if we don't like this behavior.

**Comment by sbueringer** [comment]
> > Looks like it's counting SynchronizedBeforeSuite as a spec in the Junit reporting, not sure if that's expected or not
> 
> @CecileRobertMichon I see that changed, but I haven't been able to track down where and why. I assume it's part of the new `--junit-report` implementation. It seems benign, but I can look into it more deeply if we don't like this behavior.

It would be interesting to know, but definitely not a showstopper for me.

**Review by CecileRobertMichon — APPROVED** [review]
/lgtm
/approve
/hold

feel free to remove the hold when we're ready to merge

**Comment by CecileRobertMichon** [comment]
> I see that changed, but I haven't been able to track down where and why. I assume it's part of the new --junit-report implementation. It seems benign, but I can look into it more deeply if we don't like this behavior.

definitely not a blocker, just seemed interesting

**Comment by mboersma** [comment]
/hold cancel

I think this is ready to merge. Thanks everyone for the reviews and the help!

**Comment by sbueringer** [comment]
@mboersma Thank you very much for pushing this forward! :)

**Pull request merged**
